privacyidea/api/monitoring.py

```diff
@@ -0,0 +1,99 @@
+# -*- coding: utf-8 -*-
+#
+# http://www.privacyidea.org
+#
+# 2018-08-01 Cornelius Kölbel, <cornelius.koelbel@netknights.it>
+#            Initial writeup
+#
+# This code is free software; you can redistribute it and/or
+# modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
+# License as published by the Free Software Foundation; either
+# version 3 of the License, or any later version.
+#
+# This code is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+#
+# You should have received a copy of the GNU Affero General Public
+# License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+This endpoint is used fetch monitoring/statistics data
+
+The code of this module is tested in tests/test_api_monitoring.py
+"""
+from flask import (Blueprint,
+                   request)
+from .lib.utils import getParam, send_result
+from ..lib.utils import parse_legacy_time
+from ..lib.log import log_with
+from ..lib.monitoringstats import (get_stats_keys, get_values,
+                                   get_last_value, delete_stats)
+from flask import g
+import logging
+from ..api.lib.prepolicy import prepolicy, check_base_action
+from ..lib.policy import ACTION
+
+
+log = logging.getLogger(__name__)
+
+
+monitoring_blueprint = Blueprint('monitoring_blueprint', __name__)
+
+
+@monitoring_blueprint.route('/', methods=['GET'])
+@monitoring_blueprint.route('/<stats_key>', methods=['GET'])
+@log_with(log)
+@prepolicy(check_base_action, request, ACTION.STATISTICSREAD)
+def get_statistics(stats_key=None):
+    """
+    return a list of all available statistics keys in the database if no *stats_key*
+    is specified.
+
+    If a stats_key is specified it returns the data of this key.
+    The parameters "start" and "end" can be used to specify a time window,
+    from which the statistics data should be fetched.
+    """
+    if stats_key is None:
+        stats_keys = get_stats_keys()
+        g.audit_object.log({"success": True})
+        return send_result(stats_keys)
+    else:
+        param = request.all_data
+        start = getParam(param, "start")
+        if start:
+            start = parse_legacy_time(start, return_date=True)
+        end = getParam(param, "end")
+        if end:
+            end = parse_legacy_time(end, return_date=True)
+        values = get_values(stats_key=stats_key, start_timestamp=start, end_timestamp=end)
+        g.audit_object.log({"success": True})
+        return send_result(values)
+
+
+@monitoring_blueprint.route('/<stats_key>/<start>/<end>', methods=['DELETE'])
```

Comment:
Hm, I'm by no means a REST expert, but I think specifying the start and end timestamp as part of the path is not very RESTful (because they do not point to an object). I think it would be nicer to have ``start`` and ``end`` as parameters, just like in ``get_timestamps`` above?
i.e.
```
DELETE /mykey?start=xyz&end=abc
```
... and maybe raise a ParameterError if start and end are not specified?

Reply:
I **thought** the DELETE method would not allow any parameters!?
This is why I added it this way.

Also: If you think of *each* single monitoring point, the resource would address different resources,

i.e. ``/counter1/2018-10-1/2018-10-10``  are different resources than ``/counter1/2018-11-1/2018-11-30``.

Reply:
Hm, good question. e.g. [This page](https://apihandyman.io/api-design-tips-and-tricks-getting-creating-updating-or-deleting-multiple-resources-in-one-api-call/#get-or-delete-multiple-resources) sounds like query parameters are possible for DELETE requests. I think this would be especially nice because we could use the exact same URL to get and delete monitoring points: ``GET /counter1?start=...&end=...`` and ``DELETE /counter1?start=...&end=...``.

regarding your second paragraph: That's true, but I still think that the analogy between paths and resources is not so nice here then.  Assume e.g. that there are no measurements on 2018-10-9. Then, ``/counter1/2018-10-1/2018-10-9`` and ``/counter1/2018-10-1/2018-10-10`` are different URLs, but denote the same resources.

Reply:
I think all other DELETE requests in privacyIDEA did not take any parameter yet, since they did not need one.
We do not delete audit logs.
We delete tokens, but this is ``DELETE /token/<serial>``.

And I also *think* to remember, that ``self.app.test_request_context`` does not accept ``data`` or ``params`` in case the method is DELETE.

(Honestly I am not so convinced by the URL format either, but I *think* it will not work in another way)